### PR TITLE
Scope requested resource IDs by cluster and resource kind

### PR DESCRIPTION
**Pull request by nklaassen**
For search-based access requests, resource IDs will need to be scoped by cluster and resource kind to ensure that they are unique across clusters.

I'm modifying protobuf types here but this is an early, unreleased feature I'm working on that no-one is using yet so it should be fine.

Unfortunately this now requires the CLI to work with JSON blobs:
```bash
Nics-MacBook-Pro:teletest nklaassen$ tsh request search --kind node --search one-node-iot
Kind Hostname     Labels          ID
---- ------------ --------------- ------------------------------------
node one-node-iot owner=db-admins d774a757-d0cc-4259-8022-61c8d63450fa

To request access to these resources, run
> tsh request create --resources '[{"cluster":"cluster-one","kind":"node","name":"d774a757-d0cc-4259-8022-61c8d63450fa"}]' \
    --reason <request reason>

Nics-MacBook-Pro:teletest nklaassen$ tsh request create --resources '[{"cluster":"cluster-one","kind":"node","name":"d774a757-d0cc-4259-8022-61c8d63450fa"}]'
Creating request...
Request ID: cbeabfac-5794-41f5-a757-b12d9a49cd98
Username:   nic
Roles:      db-admins
Reason:     [none]
Reviewers:  [none] (suggested)
Status:     PENDING

hint: use 'tsh login --request-id=<request-id>' to login with an approved request

Waiting for request approval...
```

This prepares for the following PR: https://github.com/gravitational/teleport/pull/12494

### CODE REVIEW

**Comment by nklaassen** [comment]
friendly ping @Tener @timothyb89

**Review by timothyb89 — APPROVED** [review]

File: api/types/events/events.proto [hunk 1]
```diff
@@ -762,9 +762,8 @@ message AccessRequestCreate {
     // variant).
     string ProposedState = 11 [ (gogoproto.jsontag) = "proposed_state,omitempty" ];
 
-    // RequestedResourceIDs is a list of resources to which access is being
-    // requested.
-    repeated string RequestedResourceIDs = 12 [ (gogoproto.jsontag) = "resource_ids,omitempty" ];
+    // RequestedResourceIDs is the set of resources to which access is being requested.
+    string RequestedResourceIDs = 12 [ (gogoproto.jsontag) = "requested_resource_ids,omitempty" ];
```
Comment (Tener) [inline api/types/events/events.proto:766]: Why don't we reference the protobuf type here instead of using embedded json?

File: lib/auth/auth.go [hunk 2]
```diff
@@ -2433,7 +2437,7 @@ func (a *Server) CreateAccessRequest(ctx context.Context, req types.AccessReques
 			Expires: req.GetAccessExpiry(),
 		},
 		Roles:                req.GetRoles(),
-		RequestedResourceIDs: req.GetRequestedResourceIDs(),
+		RequestedResourceIDs: requestedResourcesStr,
```
Comment (Tener) [inline lib/auth/auth.go:2440]: As in `.proto` file: I believe we would be better off with strongly typed fields here.

File: lib/services/access_request.go [hunk 3]
```diff
@@ -46,12 +46,12 @@ func ValidateAccessRequest(ar types.AccessRequest) error {
 
 // NewAccessRequest assembles an AccessRequest resource.
 func NewAccessRequest(user string, roles ...string) (types.AccessRequest, error) {
-	return NewAccessRequestWithResources(user, roles, nil)
+	return NewAccessRequestWithResources(user, roles, []types.ResourceID{})
```
Comment (Tener) [inline lib/services/access_request.go:49]: Can we make the code work with `nil` slice instead of empty slice?

File: tool/tsh/tsh.go [hunk 4]
```diff
@@ -1523,8 +1523,14 @@ func executeAccessRequest(cf *CLIConf, tc *client.TeleportClient) error {
 	} else {
 		roles := utils.SplitIdentifiers(cf.DesiredRoles)
 		reviewers := utils.SplitIdentifiers(cf.SuggestedReviewers)
-		resourceIDs := utils.SplitIdentifiers(cf.RequestedResourceIDs)
-		req, err = services.NewAccessRequestWithResources(cf.Username, roles, resourceIDs)
+		requestedResourceIDs := []types.ResourceID{}
+		if cf.RequestedResourceIDs != "" {
```
Comment (Tener) [inline tool/tsh/tsh.go:1527]: Can you add an example of `tsh` usage with this flag to the PR description?

File: api/types/events/events.proto (hunk 1 above)
Comment (nklaassen) [reply api/types/events/events.proto]: I brought this up on slack, ended up duplicating the type in the events package https://gravitational.slack.com/archives/C0DF0TPMY/p1652286139700439

File: lib/services/access_request.go (hunk 3 above)
Comment (nklaassen) [reply lib/services/access_request.go:50]: it does, I'm just being explicit here, I don't like seeing `nil` and having it not be completely obvious what it is

File: tool/tsh/tsh.go (hunk 4 above)
Comment (nklaassen) [reply tool/tsh/tsh.go:1527]: added

File: api/types/events/events.proto [hunk 5]
```diff
@@ -763,7 +763,18 @@ message AccessRequestCreate {
     string ProposedState = 11 [ (gogoproto.jsontag) = "proposed_state,omitempty" ];
 
     // RequestedResourceIDs is the set of resources to which access is being requested.
-    string RequestedResourceIDs = 12 [ (gogoproto.jsontag) = "requested_resource_ids,omitempty" ];
+    repeated ResourceID RequestedResourceIDs = 12
+        [ (gogoproto.jsontag) = "resource_ids,omitempty", (gogoproto.nullable) = false ];
+}
+
+// ResourceID is a unique identifier for a teleport resource.
```
Comment (Tener) [inline api/types/events/events.proto:770]: Nit: perhaps worth explaining the duplication w.r.t. the other message type.

**Review by Tener — APPROVED** [review]
I'm not entirely convinced about the proto duplication being the best choice, but I don't have strong enough arguments either, so +1 from me.

**Pull request merged**